Set minimum required standard to C++17

martyngigg opened the pull request:
**Description of work.**

Require a compiler with support for the C++17 language standard as a minimum. Note that this does not require a C++17 compliant std library implementation. For example macOS High Sierra has Apple Clang 10 (based on LLVM clang 6) that is C++17 language complete but libcxx does not have std::filesystem yet.

Despite this, access to the C++17 language features is still a win. 

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**

Builds should pass.

Fixes #26464

*This does not require release notes* because **it is maintenance.**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair commented:
It looks like we have a few deprecated call to `std::mem_fun_ref` which can be directly replaced with `std::mem_fn_ref` as it's encapsulated to the class implementation

Is this something which would be tackled in this PR or in a separate one which will block this?

martyngigg commented:
Thanks, David. I've fixed the minor `mem_fun` issues and rebased this on to of my PR to move boost forward as apparently we need that too. Would you have time to review #26480?

DavidFair commented:
Okay, I'll review that today first - thanks for the heads up

martyngigg commented:
It looks like clang is far better at actually removing stuff that is not supposed to be there. I'll compile on mac to figure the rest out.

martyngigg commented:
> Okay, I'll review that today first - thanks for the heads up

Thanks!

martyngigg commented:
And it looks like we need to upgrade tbb on Windows too :disappointed:

DavidFair commented on Framework/API/src/FileFinder.cpp:
```diff
@@ -207,8 +205,8 @@ FileFinderImpl::toInstrumentAndNumber(const std::string &hint) const {
-    std::string::const_reverse_iterator it = std::find_if(
-        hint.rbegin(), hint.rend(), std::not1(std::ptr_fun(isdigit)));
+    std::string::const_reverse_iterator it =
+        std::find_if(hint.rbegin(), hint.rend(), std::not_fn(isdigit));
```
Could we take this opportunity to replace all of the `std::not_fn` with lambdas. That way they don't have to lookup `std::not_fn` if they aren't familiar:

`std::find_if(hint.rbegin(), hint.rend(), [](const char c){return !isdigit(c);});`

martyngigg replied:
Do you not think `not_fn(isdigit)` is more readable? I have to read further with the lambda and make sure I don't miss the `!` in there.

DavidFair replied:
🤔 Good point - it's finding a balance between dev's infrequently seeing `std::not_fn` and the boilerplate of a lambda

I think I'll agree with you on this one, as once you've looked up the various memory references the STL gives you it's obvious to read

DavidFair commented on Framework/Beamline/test/ComponentInfoTest.h:
```diff
@@ -755,7 +755,7 @@ class ComponentInfoTest : public CxxTest::TestSuite {
                   .isApprox(subCompPosition - rootPosition));
 
-    const auto diffPos =
+    const Vector3d diffPos =
```
Could we add a comment explaining why `auto` is not a good idea, to save this bug creeping back in?

martyngigg replied:
:+1:

DavidFair commented on Framework/DataObjects/test/Histogram1DTest.h:
```diff
@@ -170,16 +171,22 @@ class Histogram1DTest : public CxxTest::TestSuite {
   }
   void testrangeexceptionX() {
     h.setPoints(x1);
+    MSVC_DIAG_OFF(4834)
```
Can we comment about the `nodiscard`

martyngigg replied:
:+1:

DavidFair commented on Framework/Geometry/src/Math/Acomp.cpp:
```diff
@@ -214,8 +214,8 @@ Complementary subtraction is by making A-B == A*B'
   // Sort the components of the list
-  std::for_each(Fparts.begin(), Fparts.end(), std::mem_fun_ref(&Acomp::Sort));
-  std::for_each(Gparts.begin(), Gparts.end(), std::mem_fun_ref(&Acomp::Sort));
+  std::for_each(Fparts.begin(), Fparts.end(), std::mem_fn(&Acomp::Sort));
```
Same with `sort` and a lambda

martyngigg replied:
So this is not actually sorting `Fparts`/`Gparts` (that is done below). It's calling `Acomp::Sort` on each entry in `Fparts`/`Gparts`. I'm not sure a lambda clarifies this much?

DanNixon commented:
It wasn't worth not merging for but the node about Eigen should be somewhere more generic, that rule (no `auto`) applies top all uses of Eigen.

martyngigg commented:
Yes. I just don't know where. Any suggestions?

DanNixon commented:
No idea. Maybe just a PSA in Slack. The Eigen docs regarding the issue are easy to find by searching "eigen auto".

DavidFair commented:
If it really became a problem, we could create an adapter class for Eigen which guarantees that even if a dev uses `auto` the adapter will manually coerce the type 

However, I think that just making the two or three people who actively touch the Eigen code aware would be a better investment of time
